bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/GardenaSmartImpl.java

```diff
@@ -12,562 +12,384 @@
+        try {
+            startWebsockets();
+            eventListener.onConnectionResumed();
+        } catch (Exception ex) {
+            eventListener.onConnectionLost();
+            logger.warn("Restarting GardenaSmart Webservice failed: {}, try restart in {} seconds", ex.getMessage(),
+                    RESTART_DELAY_SECONDS);
+            restartScheduledFuture = scheduler.schedule(() -> {
+                restartWebsockets();
+            }, RESTART_DELAY_SECONDS, TimeUnit.SECONDS);
+        }
+    }
 
-                allDevicesById = newDevicesById;
+    /**
+     * Sets the dataItem from the websocket event into the correct device.
+     */
+    private void handleDataItem(final DataItem dataItem) throws GardenaException {
+        final String deviceId = dataItem.getDeviceId();
+        Device device = allDevicesById.get(deviceId);
+        if (device == null && !(dataItem instanceof LocationDataItem)) {
+            device = new Device(deviceId);
+            allDevicesById.put(device.id, device);
+
+            if (initialized) {
+                scheduler.schedule(() -> {
+                    Device newDevice = allDevicesById.get(deviceId);
+                    newDevice.evaluateDeviceType();
+                    if (newDevice.deviceType != null) {
+                        eventListener.onNewDevice(newDevice);
+                    }
+                }, 3, TimeUnit.SECONDS);
+            }
+        }
 
-                for (Device deletedDevice : deletedDevices) {
-                    eventListener.onDeviceDeleted(deletedDevice);
-                }
+        if (device != null) {
+            device.setDataItem(dataItem);
+        }
+    }
 
-                for (Device newDevice : newDevices) {
-                    eventListener.onNewDevice(newDevice);
-                }
+    @Override
+    public void onWebSocketClose() {
+        restartWebsockets();
+    }
 
-                for (Device updatedDevice : updatedDevices) {
-                    eventListener.onDeviceUpdated(updatedDevice);
-                }
+    @Override
+    public void onWebSocketError(Throwable cause) {
+        restartWebsockets();
+    }
 
-            } catch (GardenaException ex) {
-                if (!connectionLost) {
-                    connectionLost = true;
-                    logger.warn("Connection lost to Gardena Smart Home with id '{}'", id);
-                    logger.trace("{}", ex.getMessage(), ex);
-                    eventListener.onConnectionLost();
+    @Override
+    public void onWebSocketMessage(String msg) {
+        try {
+            DataItem dataItem = gson.fromJson(msg, DataItem.class);
+            handleDataItem(dataItem);
+            Device device = allDevicesById.get(dataItem.getDeviceId());
+            if (device != null && device.active) {
+                devicesToNotify.add(device);
+
+                // delay the deviceUpdated event to filter multiple events for the same device dataItem property
```

Comment:
Did you read the comment? I saw several identical events with different timestamp values, even with zero as value. So I filter these unnecessary events so that the openHab elements do not change unnecessarily multiple times with wrong values.

Reply:
ok